source/extensions/http/header_validators/envoy_default/http2_header_validator.cc

```diff
@@ -0,0 +1,415 @@
+#include "source/extensions/http/header_validators/envoy_default/http2_header_validator.h"
+
+#include "source/extensions/http/header_validators/envoy_default/character_tables.h"
+
+#include "absl/container/node_hash_map.h"
+#include "absl/container/node_hash_set.h"
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Http {
+namespace HeaderValidators {
+namespace EnvoyDefault {
+
+using ::envoy::extensions::http::header_validators::envoy_default::v3::HeaderValidatorConfig;
+using ::Envoy::Http::HeaderString;
+using ::Envoy::Http::LowerCaseString;
+using ::Envoy::Http::Protocol;
+using HeaderValidatorFunction = ::Envoy::Http::HeaderValidator::HeaderEntryValidationResult (
+    Http2HeaderValidator::*)(const HeaderString&);
+
+struct Http2ResponseCodeDetailValues {
+  const std::string InvalidTE = "uhv.http2.invalid_te";
+  const std::string ConnectionHeaderSanitization = "uhv.http2.connection_header_rejected";
+};
+
+using Http2ResponseCodeDetail = ConstSingleton<Http2ResponseCodeDetailValues>;
+
+/*
+ * Header validation implementation for the Http/2 codec. This class follows guidance from
+ * several RFCS:
+ *
+ * RFC 3986 <https://datatracker.ietf.org/doc/html/rfc3986> URI Generic Syntax
+ * RFC 9110 <https://www.rfc-editor.org/rfc/rfc9110.html> HTTP Semantics
+ * RFC 9112 <https://www.rfc-editor.org/rfc/rfc9112.html> HTTP/1.1
+ * RFC 9113 <https://www.rfc-editor.org/rfc/rfc9113.html> HTTP/2
+ *
+ */
+Http2HeaderValidator::Http2HeaderValidator(const HeaderValidatorConfig& config, Protocol protocol,
+                                           StreamInfo::StreamInfo& stream_info)
+    : HeaderValidator(config, protocol, stream_info) {}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http2HeaderValidator::validateRequestHeaderEntry(const HeaderString& key,
+                                                 const HeaderString& value) {
+  static const absl::node_hash_map<absl::string_view, HeaderValidatorFunction> kHeaderValidatorMap{
+      {":method", &Http2HeaderValidator::validateMethodHeader},
+      {":authority", &Http2HeaderValidator::validateAuthorityHeader},
```

Comment:
I created a separate issue for this scenario that I think was accidentally closed: #21308. I could add support for validating each item in the `:authority` list but I think @yanavlasov wanted some community feedback prior to implementing anything.

I tested this by sending a http/2 request with an `:authority` and `Host` header that are different. It appears that both values are treated as in the parsed header map as `:authority` and both values are combined into a list:

```
[2022-09-08 09:55:46.591][312274][debug][http] [source/common/http/conn_manager_impl.cc:920]
    [C10][S11896658524638699712] request headers complete (end_stream=true):
':method', 'GET'
':path', '/'
':authority', 'localhost:10000,not-localhost'
':scheme', 'http'
'accept-encoding', 'gzip, deflate'

[2022-09-08 09:55:46.591][312274][debug][http] [source/common/http/conn_manager_impl.cc:903]
    [C10][S11896658524638699712] request end stream
[2022-09-08 09:55:46.591][312274][debug][http] [source/common/http/filter_manager.cc:905]
    [C10][S11896658524638699712] Sending local reply with details uhv.invalid_host
```

In this example, I performed an HTTP/2 request with `:authority = "localhost:10000"` and `Host = "not-localhost"`. UHV detects an error in the host and returns a 400 local reply with details `uhv.invalid_host`.

Reply:
I think this is out of scope for this PR, since aliasing `Host` and `:authority` headers is the behavior of the header map, not header map validator.

@diannahu if you feel like it is necessary to have separate `Host` and `:authority` headers, please file an Issue on GitHub.

Reply:
I filed https://github.com/envoyproxy/envoy/issues/23054. Thanks!